### ISSUE TITLE
IDE0039 recommendation actually makes the code less efficient

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn#26786]
**Version Used**: 
VisualStudio.15.IntPreview/15.8.0-pre.1.1+27705.2000.d15.8

**Steps to Reproduce**:
Add this code into a project in the IDE:
```C#
using System;

class Program
{
    static void Main()
    {
        Action a = () => { };
        for (int i = 0; i < 1000; i++)
        {
            Method(a);
        }
    }

    static void Method(Action a) { }
}
```
The IDE then flags the delegate and says to change it from a delegate to a local function:
![image](https://user-images.githubusercontent.com/2642209/39902856-37c9ad24-549e-11e8-8d99-a60d4993f95a.png)

The problem with this is that in a case like this where I actually need to use it as a delegate, it results in worse code.  The original code is compiled to the equivalent of this:
```C#
private static void Main()
{
    Action a = (<>c.<>9__0_0 != null) ? <>c.<>9__0_0 : (<>c.<>9__0_0 = new Action(this.<Main>b__0_0));
    for (int i = 0; i < 1000; i++)
    {
        Method(a);
    }
}
```
whereas if I apply the recommended fix to end up with the code:
```C#
using System;

class Program
{
    static void Main()
    {
        void a() { }
        for (int i = 0; i < 1000; i++)
        {
            Method(a);
        }
    }

    static void Method(Action a) { }
}
```
this ends up compiling to code equivalent to this:
```C#
private static void Main()
{
    for (int i = 0; i < 1000; i++)
    {
        Method(new Action(<>c.<>9.<Main>g__a|0_0));
    }
}
```
Now rather than one delegate allocation for all 1000 iterations, I end up with 1000 delegate allocations.

**Comment by CyrusNajmabadi** [comment]
Interesting.  Seems unfortunate that we don't create a single instance to pass along.  That seems to go along with the discussions that have been had about caching a normal method-passed-to-a-delegate.  @agocke Is this something that could be done here?

**Comment by agocke** [comment]
Yes, we should probably use delegate caching for local functions.

**Comment by svick** [comment]
The related issue for caching a delegate for method group conversion is https://github.com/dotnet/roslyn/issues/5835.

**Comment by sharwell** [comment]
> Yes, we should probably use delegate caching for local functions.

I would very much prefer this solution to disabling the analyzer. Also, for many end-user applications, local functions offer improved readability with negligible observed performance overhead.

**Comment by stephentoub** [comment]
> I would very much prefer this solution to disabling the analyzer.

We do the delegate caching (for all method groups, not just local functions).  But I'd also like to see this analyzer improved to understand that if the only consumer(s) of the local function is via conversions to delegates, the warning should not fire: in my opinion, such a change offers decreased understability, not improved, nevermind the perf implications.

**Comment by svick** [comment]
@stephentoub 

It's not a Warning, it's [Info](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.diagnosticseverity?view=roslyn-dotnet) ("Information that does not indicate a problem (i.e. not prescriptive)."). Would it help if it was changed from a code fix to a code refactoring (similar to e.g. the "for to foreach" and "foreach to for" refactorings), which would make it less prominent and appear less as something that you should do?

**Comment by CyrusNajmabadi** [comment]
> To my knowledge I've never changed any defaults here, but maybe if it now defaults to "suggestion" it's somehow retained a stale setting on my machine.

Yeah.  Roslyn has never shipped anything at the 'error' level.  So something has gone either *very* wrong, or it was something you did :D    I've never seen anyone report anything about our fixes evern being above the 'suggestion' level.  We're actually super wary of doing that as we really just want to surface *potential* improvements, but not actually scare people with warnings or errors.

**Comment by CyrusNajmabadi** [comment]
> and it's in the eye of the beholder whether a local function is better in that case... for me, it's definitely not.

Sure.  That's why it's just "info".  Literally "information" :)   The intent is to make you aware that you *could* make this change, but in no way expect that you *should* make this change.

**Comment by stephentoub** [comment]
> That's why it's just "info". Literally "information"

It's called "Suggestion"; that's more than "information". :) So the IDE here is, by default, "suggesting" I do something that makes my code less efficient.

**Comment by CyrusNajmabadi** [comment]
> and it's in the eye of the beholder whether a local function is better in that case... for me, it's definitely not.

Well, that's the case with a lot of suggestions :)   I mean, some people may not want collection/object initializers and whatnot.

That said, we could always consider an additional option to potentially control this subbehavior.  Not sure if it would ever be in teh VS ui.  But it might be something we have in .editorconfig.

**Comment by CyrusNajmabadi** [comment]
> It's called "Suggestion"; that's more than "information". :) 

Damn, caught that before it was removed.  :)  I was remembering the orignal designs, when they were called 'info' :D

**Comment by CyrusNajmabadi** [comment]
> So the IDE here is, by default, "suggesting" I do something that makes my code less efficient.

well, let's fix the efficiency problem :)   It's not good that we have this nice new feature and it has this pitfall :D

**Comment by stephentoub** [comment]
> let's fix the efficiency problem

I agree; that issue has been open for two and a half years, though.

**Comment by agocke** [comment]
Well, it is a language change. The language specifies that a new delegate is created.
